### ISSUE TITLE
Investigate ways to reduce browser test duration

### ISSUE THREAD

**Issue by gwendolyngoetz** [civiform/civiform#4111]
The time it takes to run the browser tests has been growing for a while. Seattle recently increased the timeout from 50 to 70 minutes since it was causing the probers to fail.

We should look to see if any tests are no longer needed.

Some other ideas include:
* Separating accessibility tests into a separate step
* See if playwright can utilize gpu acceleration
* Separating out tests that are related to testing with a feature flag on (when there is a corresponding test with it off)

This data from this graph was pulled from successful Seattle staging jobs going back to August 2022 and only for the run probers step.

![image](https://user-images.githubusercontent.com/195162/214933956-da6555b8-6e2a-4d4e-bd2d-9f08850e3dfa.png)

**Comment by jessbumblebee** [comment]
A few additional ideas:

Optimizing utility methods
A lot of the utility methods (e.g. program_admin.ts) are optimised for readability/re-usability rather than speed. For example, they always return to the main page like in the link below and it is not necessay in every case. If every tests adds 2-3 unnecessary clicks, it may impact performance significantly
https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/admin_programs.ts?L103&subtree=true

Rewrite based on user flows
At the moment browser tests are written based on the feature someone adds, but often several features require almost the exact same steps to test them. we could re-assess if there are tests that would be more efficient if they were merged.

BeforeAll:
Sounds to me like you can optimise by moving steps into this to do it only once before tests are started 
https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L222:3&subtree=true

describes
If brining up each test file causes a high cost, we could migrate to having more cases of multiple "describes" in each test. This is something Shane briefly mentioned before he left.
https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/civiform_admin_program_statuses.test.ts?L19&subtree=true

long term maintenance
Knowledge about which code come at what cost would be very valuable information to help people do the right thing in the long run. For example adding a completely new test file vs adding an additional test to a existing file; adding modular tests vs doing larger single tests, screenshots vs clicking, what makes a click expensive etc.
We could establish knowledge and best practices based on it.

**Comment by nbeloglazov** [comment]
From my investigation what made tests slower is DB initialization. Usually test needs to prepare data and it's done in before/beforeAll using clicking through UI. And that part can be optimized, but couldn't think of really nice way to do it. Here are my thoughts on that issue: https://github.com/civiform/civiform/issues/2891#issuecomment-1235813976

I think if we could prepare data programmatically and create by sending a single JSON request - that would help a lot, but it requires creating TS API for creating programs/questions and that requires syncing that API with server-side java code.

**Comment by shanemc-goog** [comment]
FWIW it looks like server time is not the main issue.

I hacked up the /metrics controller to report average time by handler in #4113 

And the total time in the server is a minority of the total test run time.
[data
](https://docs.google.com/spreadsheets/d/1Q-jLFo3RjUdkGi3LNtAW0q5FzJI8fUgxwk2eB17wpFY/edit#gid=0)

All tests: Wall time: 1355s   Server time: 142s
src/question_lifecycle.test.ts: Wall time: 165s  Server time 29s

```
Test Suites: 1 passed, 1 total
Tests:       20 passed, 20 total
Snapshots:   39 passed, 39 total
Time:        165.343 s
```

**Comment by shanemc-goog** [comment]
Screenshots also look to not be the core issue, though they're likely cumulatively a non-insignificant amount of time.

Profiling interaction with the browser would be my next thought, though unsure how.  EG inspecting the page to find elements and filling out the forms etc.

**Comment by shanemc-goog** [comment]
A decent culprite is the wait for the question bank animation to open/close, we **always** wait 1s to evaluate the animation completing

https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/admin_programs.ts?L357&subtree=true

Lowering it to 250ms made src/question_lifecycle.test.ts run in 131s as compared to 165s (from the post above) for me.  The test file has 14 question bank interactions and each one waits the 1s twice for open and close.

Note that that is also an arbitrary wait and not for an end result in the DOM

**Comment by nb1701** [comment]
Needs more discussion on the right way forward.  In the short term, should get better when we clean up feature flags.  But need a better long term solution.  Maybe flag certain tests to run in the hermetic environment where we can parallelize them, and flag some tests as critical to run in the prober environment.

**Comment by rockycodes** [comment]
From discussion at 7/13/26 [eng weekly](https://docs.google.com/spreadsheets/d/1XXEmV5zxzZsIMtgGwGjZrqgjsa6ic3hRZHuwXZg0eI4/edit?gid=36363296#gid=36363296), we want to migrate browser tests that aren't explicitly testing the admin flow, to use seeding logic to populate programs and questions where logical. If specific programs or questions are needed for a test, they should still be created through the admin UI. See example PRs of how to refactor to use seeding logic:
https://github.com/civiform/civiform/pull/13612
https://github.com/civiform/civiform/pull/13621
https://github.com/civiform/civiform/pull/13622
